src/MessagePack/Formatters/FrozenCollectionFormatters.cs

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) All contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#if NET8_0_OR_GREATER
+
+using System.Collections.Frozen;
+using System.Collections.Generic;
+using MessagePack.Formatters;
+
+#pragma warning disable SA1402 // File may only contain a single type
+#pragma warning disable SA1649 // File name should match first type name
+
+namespace MessagePack.ImmutableCollection
+{
+    public sealed class FrozenDictionaryFormatter<TKey, TValue> : IMessagePackFormatter<FrozenDictionary<TKey, TValue>?>
+        where TKey : notnull
+    {
+        private readonly IEqualityComparer<TKey>? comparer;
+
+        public FrozenDictionaryFormatter()
+        {
+            comparer = default;
+        }
+
+        public FrozenDictionaryFormatter(IEqualityComparer<TKey>? comparer)
+        {
+            this.comparer = comparer;
+        }
+
+        public void Serialize(ref MessagePackWriter writer, FrozenDictionary<TKey, TValue>? value, MessagePackSerializerOptions options)
+        {
+            if (value is null)
+            {
+                writer.WriteNil();
+                return;
+            }
+
+            IFormatterResolver resolver = options.Resolver;
+            IMessagePackFormatter<TKey> keyFormatter = resolver.GetFormatterWithVerify<TKey>();
+            IMessagePackFormatter<TValue> valueFormatter = resolver.GetFormatterWithVerify<TValue>();
+
+            // https://github.com/dotnet/runtime/blob/4c500699b938d53993b928b93543b8dbe68f69aa/src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/FrozenHashTable.cs#L134C2-L134C2
+            // FrozenDictionary.Count uses FrozenHashTable's Count property which is O(1).
+            var count = value.Count;
+            writer.WriteMapHeader(count);
+            if (count == 0)
+            {
+                return;
+            }
+
+            FrozenDictionary<TKey, TValue>.Enumerator e = value.GetEnumerator();
+            while (e.MoveNext())
+            {
```

Comment:
Why not use `foreach` here?

Reply:
I think the reason I did it that way was because I wanted to check whether the enumerator was a struct or a class, so I wrote it that way.
I change all the enumerator-patterns to foreach.